src/transformers/modeling_blenderbot.py

```diff
@@ -0,0 +1,582 @@
+            'encoder_attn': final_enc_attn_incr_state
+        }
+        return input, new_incremental_state
+
+
+class BlenderbotDecoder(nn.Module):
+    """
+    implements the decoder part of the blender bot
+    """
+    def __init__(self, n_heads, embedding_size, ffn_size, n_layers, vocabulary_size, embedding=None,
+                 attention_dropout=0.0, dropout=0.0, relu_dropout=0.0, embedding_scale=True,
+                 learn_positional_embeddings=False, padding_idx=0, n_positions = 1024, n_segments=0,
+                 activation='relu'):
+        """
+
+        :param n_heads: number of multi attention heads in the decoder
+        :param embedding_size: embedding size
+        :param ffn_size:  size of hidden layers in the FFN
+        :param n_layers:  number of layers
+        :param vocabulary_size: size of the vocabulary
+        :param embedding:  embedding matrix
+        :param attention_dropout: dropout after the multihead attention
+        :param dropout:  dropout around embeddings
+        :param relu_dropout: dropout after Relu
+        :param embedding_scale:  scale embeddings
+        :param learn_positional_encodings: If off, sinusoidal embeddings are used. If on, position embeddings are learned from scratch.
+        :param padding_idx:padding index
+        :param n_positions:size of the position embeddings
+        :param n_segments:
+        :param activation: activation function to use
+        """
+        super().__init__()
+        self.n_heads = n_heads
+        self.n_layers = n_layers
+        self.ffn_size = ffn_size
+        self.embedding_size = embedding_size
+        self.vocabulary_size = vocabulary_size
+        self.dropout = nn.Dropout(p=dropout)
+        self.embedding_scale = embedding_scale
+        self.n_positions = n_positions
+        self.n_segment = n_segments
+        self.embeddings = embedding
+        assert self.embedding_size % self.n_heads == 0, "embedding size must be a multiple of n_heads"
+        self.positional_embeddings = nn.Embedding(self.n_positions, self.embedding_size)
+        if not learn_positional_embeddings:
+            positional_coding(self.n_positions, self.embedding_size, self.positional_embeddings)
+        else:
+            nn.init.normal_(self.positional_embeddings.weight, 0, self.embedding_size**(-0.5))
+        self.layers = nn.ModuleList()
+        for i in range(self.n_layers):
+            self.layers.append(BlenderbotDecoderLayer(self.n_heads, self.embedding_size, self.ffn_size,
+                                                      activation, dropout, attention_dropout, relu_dropout))
+
+    def forward(self, input, encoder_state, incremental_state=None):
+        encoder_output, encoder_mask = encoder_state
+        seq_len = input.size(1)
+        positions = input.new(seq_len).long()
+        positions = torch.arange(seq_len, out=positions).unsqueeze(0)
+        if incremental_state:
+            input = input[:, -1:]
```

Comment:
No automatic cutting of the `input`. I think I was the one who introduces that not so favorable design choice, but after discussion with @thomwolf and @LysandreJik we decided to not do this. Discussion is here: https://github.com/huggingface/transformers/issues/4368#issuecomment-630244541